### PR TITLE
Procedure to invalidate directory list cache

### PR DESCRIPTION
**Pull request by agrawalreetika**
Test plan - (Please fill in how you tested your changes)

Please make sure your submission complies with our [Development](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#development), [Formatting](https://github.com/prestodb/presto/wiki/Presto-Development-Guidelines#formatting), and [Commit Message](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#commit-formatting-and-pull-requests) guidelines. Don't forget to follow our [attribution guidelines](https://github.com/prestodb/presto/wiki/Review-and-Commit-guidelines#attribution) for any code copied from other projects.

Fill in the release notes towards the bottom of the PR description.
See [Release Notes Guidelines](https://github.com/prestodb/presto/wiki/Release-Notes-Guidelines) for details.

```
== RELEASE NOTES ==


Hive Changes
* Introduce system procedure to invalidate directory list cache in Hive Catalog
```

### CODE REVIEW

**Comment by nmahadevuni** [comment]
@agrawalreetika Instead of single directory, we may want to clear the whole cache too. May be make the directoryPath optional, and if its not provided, then clear all cache. Also, please change the procedure name to something more intuitive.

If there are multiple hive catalogs, how can we invalidate the cache from one particular catalog?

`system.invalidate_file_list_cache(Optional<String> directoryPath)`

**Comment by agrawalreetika** [comment]
> @agrawalreetika Instead of single directory, we may want to clear the whole cache too. May be make the directoryPath optional, and if its not provided, then clear all cache. Also, please change the procedure name to something more intuitive.
> 
> If there are multiple hive catalogs, how can we invalidate the cache from one particular catalog?
> 
> `system.invalidate_file_list_cache(Optional<String> directoryPath)`

Hi @nmahadevuni, Thanks for your comment.
I have added the directory path as optional to allow full directory list cache invalidation.
Just like any other procedure in Presto we will have to provide catalog name while making procedure call on specific catalog. (ref - https://prestodb.io/docs/current/connector/hive.html#procedures)

File: presto-hive/src/main/java/com/facebook/presto/hive/CacheInvalidationProcedure.java [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+import java.util.List;
+import java.util.Optional;
+
+import static com.facebook.presto.common.block.MethodHandleUtil.methodHandle;
+import static com.facebook.presto.common.type.StandardTypes.VARCHAR;
+import static com.facebook.presto.hive.HiveSessionProperties.isUseListDirectoryCache;
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_PROCEDURE_ARGUMENT;
+import static java.util.Objects.requireNonNull;
+
+public class CacheInvalidationProcedure
```
Comment (imjalpreet) [inline presto-hive/src/main/java/com/facebook/presto/hive/CacheInvalidationProcedure.java:38]: nit: I think we can rename this class to `FileListCacheInvalidationProcedure` or `DirectoryListCacheInvalidationProcedure` since we might add more procedures for other caches.

File: presto-hive/src/main/java/com/facebook/presto/hive/CacheInvalidationProcedure.java [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+                else {
+                    throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Given directory path is not cached : " + directoryPath);
+                }
+            }
+            else {
+                cachingDirectoryLister.flushCache();
+            }
+        }
+        else {
+            throw new PrestoException(INVALID_PROCEDURE_ARGUMENT, "Directory list cache is not enabled on catalog");
```
Comment (imjalpreet) [inline presto-hive/src/main/java/com/facebook/presto/hive/CacheInvalidationProcedure.java:93]: nit: Directory list cache is not enabled on this catalog.

Also, imo we should probably have the error messages and procedure naming in sync. We can decide whether we want to call it File List Cache or Directory List Cache.

**Review by imjalpreet — CHANGES_REQUESTED** [review]
@agrawalreetika thanks for the PR, overall LGTM. A few minor nits.

File: presto-hive/src/main/java/com/facebook/presto/hive/CacheInvalidationProcedure.java (hunk 1 above)
Comment (agrawalreetika) [reply presto-hive/src/main/java/com/facebook/presto/hive/CacheInvalidationProcedure.java:38]: Yes make sense

**Review by imjalpreet — APPROVED** [review]
LGTM 👍🏼

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java [hunk 3]
```diff
@@ -134,6 +134,7 @@ public void configure(Binder binder)
         binder.bind(PrestoS3ClientFactory.class).in(Scopes.SINGLETON);
 
         binder.bind(DirectoryLister.class).annotatedWith(ForCachingDirectoryLister.class).to(HadoopDirectoryLister.class).in(Scopes.SINGLETON);
+        binder.bind(CachingDirectoryLister.class).in(Scopes.SINGLETON);
```
Comment (tdcmeehan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java:137]: Why is this needed?

File: presto-hive/src/main/java/com/facebook/presto/hive/DirectoryListCacheInvalidationProcedure.java [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+    }
+
+    public void directoryListCacheInvalidation(ConnectorSession session, String directoryPath)
+    {
+        try (ThreadContextClassLoader ignored = new ThreadContextClassLoader(getClass().getClassLoader())) {
+            doInvalidateDirectoryListCache(session, Optional.ofNullable(directoryPath));
+        }
+    }
+
+    private void doInvalidateDirectoryListCache(ConnectorSession session, Optional<String> directoryPath)
```
Comment (tdcmeehan) [inline presto-hive/src/main/java/com/facebook/presto/hive/DirectoryListCacheInvalidationProcedure.java:73]: Can we improve the enapsulation of `CachingDirectoryLister` by moving this method to be inside of it?

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java (hunk 3 above)
Comment (agrawalreetika) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java:137]: To Inject CachingDirectoryLister in DirectoryListCacheInvalidationProcedure, Explicit binding is required for this

Comment (tdcmeehan) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java:137]: Could we inject a `DirectoryLister` and do a check if it's an instance of `CachingDirectoryLister`?

Comment (agrawalreetika) [reply presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java:137]: Yes, I think we could do that

File: presto-hive/src/main/java/com/facebook/presto/hive/DirectoryListCacheInvalidationProcedure.java (hunk 4 above)
Comment (agrawalreetika) [reply presto-hive/src/main/java/com/facebook/presto/hive/DirectoryListCacheInvalidationProcedure.java:73]: @tdcmeehan Since we are going to use this for the Directory list invalidation procedure, which is why I thought of keeping it here. Do you think we should move this to CachingDirectoryLister?

I have added tests.

File: presto-hive/src/main/java/com/facebook/presto/hive/CachingDirectoryLister.java [hunk 5]
```diff
@@ -42,7 +42,7 @@
 public class CachingDirectoryLister
         implements DirectoryLister
 {
-    private final Cache<Path, List<HiveFileInfo>> cache;
+    protected final Cache<Path, List<HiveFileInfo>> cache;
```
Comment (imjalpreet) [inline presto-hive/src/main/java/com/facebook/presto/hive/CachingDirectoryLister.java:45]: nit: Can we keep the protected variables together?

File: presto-hive/src/main/java/com/facebook/presto/hive/DirectoryListCacheInvalidationProcedure.java (hunk 4 above)
Comment (tdcmeehan) [reply presto-hive/src/main/java/com/facebook/presto/hive/DirectoryListCacheInvalidationProcedure.java:73]: I am in favor of keeping this inside of `CachingDirectoryLister` for improved encapsulation.

File: presto-hive/src/test/java/com/facebook/presto/hive/TestDirectoryListCacheInvalidation.java [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        assertEquals(result.getRowCount(), 1);
+
+        long hitCount = (long) result.getMaterializedRows().get(0).getField(0);
+        long missCount = (long) result.getMaterializedRows().get(0).getField(1);
+
+        assertEquals(hitCount, 1);
+        assertEquals(missCount, 1);
+
+        // Invalidate directory list cache
+        assertQuerySucceeds("CALL hive.system.invalidate_directory_list_cache()");
```
Comment (mrmaddx2) [inline presto-hive/src/test/java/com/facebook/presto/hive/TestDirectoryListCacheInvalidation.java:90]: would be nice to also test parametrized version.

File: presto-docs/src/main/sphinx/connector/hive.rst [hunk 7]
```diff
@@ -753,6 +753,14 @@ The following procedures are available:
     file system paths to use lowercase (e.g. ``col_x=SomeValue``). Partitions on the file system
     not conforming to this convention are ignored, unless the argument is set to ``false``.
 
+* ``system.invalidate_directory_list_cache()``
+
+    Flush full directory list cache.
+
+* ``system.invalidate_directory_list_cache(directory_path)``
```
Comment (mrmaddx2) [inline presto-docs/src/main/sphinx/connector/hive.rst:760]: what is the directory_path? How can i build it knowing eg. only a table name in metastore? Could you please provide examples?

Comment (agrawalreetika) [reply presto-docs/src/main/sphinx/connector/hive.rst:760]: @mrmaddx2 Here directory_path refers to the path which we would like to exclude from the directory list cache. Right now current procedure doesn't allow us to specify `table name` directly, may be we can add that as an extension if needed in a different procedure call.

Comment (mrmaddx2) [reply presto-docs/src/main/sphinx/connector/hive.rst:760]: Could you please provide an example?

Comment (agrawalreetika) [reply presto-docs/src/main/sphinx/connector/hive.rst:760]: Sure here ,
`CALL <catalog-name>.system.invalidate_directory_list_cache('<directory_path>')`

If your Hive catalog is called hive.
**Ex :** CALL hive.system.invalidate_directory_list_cache('s3://testTable/partition1')

**Review by steveburnett — APPROVED** [review]
LGTM! (docs)
Local build of docs, new additions look good.

**Review by tdcmeehan — APPROVED** [review]

**Comment by steveburnett** [comment]
Failing on one test (test / test (:presto-hive) (pull_request), and no option to re-run failed tests.

**Review by steveburnett — APPROVED** [review]
LGTM! (docs) 

Pull updated branch, local build of docs.

File: presto-hive/src/main/java/com/facebook/presto/hive/CachingDirectoryLister.java [hunk 8]
```diff
@@ -35,17 +36,17 @@
...
 import static com.google.common.collect.ImmutableSet.toImmutableSet;
 import static java.util.Objects.requireNonNull;
 
 public class CachingDirectoryLister
         implements DirectoryLister
 {
-    private final Cache<Path, List<HiveFileInfo>> cache;
-    private final CachedTableChecker cachedTableChecker;
-
+    protected final Cache<Path, List<HiveFileInfo>> cache;
```
Comment (tdcmeehan) [inline presto-hive/src/main/java/com/facebook/presto/hive/CachingDirectoryLister.java:47]: Can this be private now?

**Review by tdcmeehan — CHANGES_REQUESTED** [review]

File: presto-hive/src/main/java/com/facebook/presto/hive/CachingDirectoryLister.java [hunk 9]
```diff
@@ -35,18 +36,18 @@
...
 
+import static com.facebook.presto.spi.StandardErrorCode.INVALID_PROCEDURE_ARGUMENT;
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.collect.ImmutableSet.toImmutableSet;
 import static java.util.Objects.requireNonNull;
 
 public class CachingDirectoryLister
         implements DirectoryLister
 {
+    protected final DirectoryLister delegate;
```
Comment (tdcmeehan) [inline presto-hive/src/main/java/com/facebook/presto/hive/CachingDirectoryLister.java:47]: It looks like this is still protected

**Comment by agrawalreetika** [comment]
@tdcmeehan I have been facing this issue in `product-tests-basic-environment` [here](https://github.com/prestodb/presto/actions/runs/8079587262/job/22074302792?pr=19821) but not finding the details around what's causing this.
Are there any more logs that get captured as part of workflow run?

**Comment by tdcmeehan** [comment]
@agrawalreetika why do we need new product tests for this feature?  Can't it run in an existing suite?

**Comment by agrawalreetika** [comment]
> @agrawalreetika why do we need new product tests for this feature? Can't it run in an existing suite?

@tdcmeehan I added tests to the existing suite, but looks like `product-tests-basic-environment` is having same issue [here](https://github.com/prestodb/presto/actions/runs/8085771312/job/22094044940?pr=19821)

**Comment by agrawalreetika** [comment]
@tdcmeehan & @steveburnett all the tests are green, this is ready for final review.

**Review by tdcmeehan — APPROVED** [review]
